hooks/post_apply.py

```diff
@@ -0,0 +1,118 @@
+
+from external_resources_io.input import (
+    parse_model,
+    read_input_from_file,
+)
+from external_resources_io.terraform import (
+    Action,
+    TerraformJsonPlanParser,
+)
+
+from er_aws_elasticache.app_interface_input import AppInterfaceInput
+from hooks_lib import ServiceUpdatesManager
+from hooks_lib.service_updates import ServiceUpdate
+
+logging.basicConfig(level=logging.INFO, format="%(levelname)s: %(message)s")
+logging.getLogger("botocore").setLevel(logging.ERROR)
+logger = logging.getLogger(__name__)
+
+
+def terraform_changes(plan: TerraformJsonPlanParser) -> bool:
+    """Check if there are any terraform changes"""
+    return bool([
+        c
+        for c in plan.plan.resource_changes
+        if c.change and c.change.actions != [Action.ActionNoop]
+    ])
+
+
+def default_cooldown(environment: str) -> int:
+    """Calculate the cooldown period based on the environment name."""
+    name = environment.lower().strip()
+    match name:
+        case _ if "production" in name:
+            default = 14
+        case _ if "staging" in name or "stage" in name:
+            default = 7
+        case _:
+            default = 5
+
+    return default
+
+
+def dry_run_mode(service_updates: list[ServiceUpdate], *, changes: bool) -> None:
+    """Handle dry-run mode."""
+    logger.info("Service updates available:")
+    for su in service_updates:
+        logger.info(
+            f"Name={su.name} Release Date={su.release_date:%Y-%m-%d} Severity={su.severity}"
+        )
+
+    if changes:
+        logger.error(
+            "Some service updates not applied (yet) and resource changes detected. Aborting."
+        )
+        sys.exit(1)
+
+
+def main(
+    plan: TerraformJsonPlanParser,
+    app_interface_input: AppInterfaceInput,
+    *,
+    dry_run: bool,
+) -> None:
+    """Ensure that no service updates are in progress."""
+    if not app_interface_input.data.service_updates_enabled:
+        logger.info("Automatic service updates are disabled.")
+        return
+
+    sumgr = ServiceUpdatesManager(
+        app_interface_input.data.replication_group_id, app_interface_input.data.region
+    )
+
+    service_updates = sumgr.service_updates(
+        severities=app_interface_input.data.service_updates_severities,
+        released_before=dt.now(tz=UTC)
+        - timedelta(
+            days=app_interface_input.data.service_updates_cooldown_days
+            if app_interface_input.data.service_updates_cooldown_days is not None
+            else default_cooldown(app_interface_input.data.environment)
+        ),
+    )
+
+    if not service_updates:
+        # No service updates available
+        return
+
+    changes = terraform_changes(plan)
+    if dry_run:
+        dry_run_mode(service_updates, changes=changes)
+        return
+
+    if changes:
+        # do not do anything if there are resource changes
+        logger.info("Resource changes detected. Skipping any pending service updates.")
+        return
```

Comment:
should we do this check at the beginning of `main` to save aws calls when there are `changes`?

always abort first if any changes detected, then `dry_run_mode` only care about `service_updates`.